File: src/la_chona_dance.py
#!/usr/bin/env python

## Import what we need
import copy
import actionlib
import rospy

from math import sin, cos
from moveit_python import (MoveGroupInterface,
                           PlanningSceneInterface,
                           PickPlaceInterface)
from moveit_python.geometry import rotate_pose_msg_by_euler_angles

## Import from message types
from control_msgs.msg import FollowJointTrajectoryAction,FollowJointTrajectoryFeedback, FollowJointTrajectoryGoal
from control_msgs.msg import PointHeadAction,PointHeadFeedback, PointHeadGoal
from control_msgs.msg import GripperCommandGoal, GripperCommandAction, GripperCommandFeedback

from grasping_msgs.msg import FindGraspableObjectsAction, FindGraspableObjectsGoal
from geometry_msgs.msg import PoseStamped, Twist, Point
from moveit_msgs.msg import PlaceLocation, MoveItErrorCodes
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint


class FollowTrajectoryClient(object):
    def __init__(self):
        ## Setup action client that plans around objects. This is ideal for getting robot to move to it's initial position
        rospy.loginfo("Waiting for MoveIt...")
        self.move_group = MoveGroupInterface("arm_with_torso", "base_link")
        rospy.loginfo("...connected")

        ## Setup action client that allows us to use the feedback callback
        self.fast_client = actionlib.SimpleActionClient("/arm_with_torso_controller/follow_joint_trajectory" ,
                                                   FollowJointTrajectoryAction)
        rospy.loginfo("Waiting for Joint trajectory...")
        self.fast_client.wait_for_server()
        rospy.loginfo("...connected")

        ## Set the names of the joints
        self.joint_names = ["torso_lift_joint", "shoulder_pan_joint", "shoulder_lift_joint", "upperarm_roll_joint",
                  "elbow_flex_joint", "forearm_roll_joint", "wrist_flex_joint", "wrist_roll_joint"]

        ## Set up action client for gripper
        rospy.loginfo("Waiting for Gripper client")
        self.gripper_client = actionlib.SimpleActionClient('gripper_controller/gripper_action', GripperCommandAction)
        self.gripper_client.wait_for_server()
        rospy.loginfo("...connected")

        ## gripper params
        self.gripper_closed_pos = 0  # The position for a fully-closed gripper (meters).
        self.gripper_open_pos = 0.10  # The position for a fully-open gripper (meters).

        ## Padding does not work (especially for self collisions)
        ## So we are adding a box above the base of the robot
        self.scene = PlanningSceneInterface("base_link")
        self.scene.removeCollisionObject("keepout")
        self.scene.addBox("keepout", 0.2, 0.5, 0.05, 0.15, 0.0, 0.375)

        self.base_action = FootWork()
        self.head_action = PointHeadClient()

    ## Function that plans and moves fetch around the "keepout" object.
    ## Function takes both the arm_and_torso joint positions and velocity arguments.
    def safe_move_to(self, positions, velocity=1):
        ## Execute motion with the moveToJointPosition function.
        ## while not rospy.is_shutdown():
        result = self.move_group.moveToJointPosition(self.joint_names,
                                                 positions,
                                                 0.0,
                                                 max_velocity_scaling_factor=velocity)

    ## This function allows the fecth to move quicker.
    ## WARNING: This does not plan around objects.
    def fast_move_to(self,positions, duration = 1, base_motion = None, head_motion = None):
        self.base_motion = base_motion
        self.head_motion = head_motion

        if len(self.joint_names) != len(positions):
            print("Invalid trajectory position")
            return False
        trajectory = JointTrajectory()
        trajectory.joint_names = self.joint_names
        trajectory.points.append(JointTrajectoryPoint())
        trajectory.points[0].positions = positions
        trajectory.points[0].velocities = [0.0 for _ in positions]
        trajectory.points[0].accelerations = [0.0 for _ in positions]
        trajectory.points[0].time_from_start = rospy.Duration(duration)
        follow_goal = FollowJointTrajectoryGoal()
        follow_goal.trajectory = trajectory

        self.fast_client.send_goal(follow_goal,feedback_cb=self.feedback_callback)
        self.fast_client.wait_for_result()


    def feedback_callback(self,feedback):
        if self.base_motion == "Forward":
            self.base_action.move_forward(1)

        elif self.base_motion == "Backward":
            self.base_action.move_backward(1)

        elif self.base_motion == "Left_turn":
            self.base_action.left_turn(1)

        elif self.base_motion == "Right_turn":
            self.base_action.right_turn(1)

        elif self.base_motion == "Wide_turn":
            self.base_action.wide_turn(1)

        if self.head_motion == "move":
            self.head_action.look_at(0, 0 ,0, frame = "gripper_link", duration = 1)

        else:
            pass

    def open_gripper(self):
        ## Functions opens Grippers
        goal = GripperCommandGoal()
        goal.command.position = self.gripper_open_pos
        self.gripper_client.send_goal(goal)
        self.gripper_client.wait_for_result()


    def close_gripper(self):
        ## Functions closses Grippers
        goal = GripperCommandGoal()
        goal.command.position = self.gripper_closed_pos
        self.gripper_client.send_goal(goal)
        self.gripper_client.wait_for_result()

## Point the head using controller
class PointHeadClient(object):

    def __init__(self):
        ## Setup action client for the head movement
        self.client = actionlib.SimpleActionClient("head_controller/point_head", PointHeadAction)
        rospy.loginfo("Waiting for head_controller...")
        self.client.wait_for_server()

    def look_at(self, x, y, z, frame = "base_link", duration=1.0):
        goal = PointHeadGoal()
        goal.target.header.stamp = rospy.Time.now()
        goal.target.header.frame_id = frame
        goal.target.point.x = x
        goal.target.point.y = y
        goal.target.point.z = z
        goal.min_duration = rospy.Duration(duration)
        self.client.send_goal(goal)
        self.client.wait_for_result()



class FootWork(object):

    def __init__(self):
        ## Initialize publisher.The Fetch will listen for Twist messages on the cmd_vel topic.
        self.pub = rospy.Publisher('cmd_vel', Twist, queue_size=10)

        ## Set up a twist message
        self.twist = Twist()
        self.twist.linear.x = 0.0
        self.twist.linear.y = 0.0
        self.twist.linear.z = 0.0
        self.twist.angular.x = 0.0
        self.twist.angular.y = 0.0
        self.twist.angular.z = 0.0

        ## Limit the publication rate.
        self.rate = rospy.Rate(10)

    def wide_turn(self, iter):
        ## Set angular rotation around z access to 1 (turn left/CCW)
        self.twist.angular.z = -1
        self.twist.linear.x = 0.6

        for i in range(iter):
            self.pub.publish(self.twist)
            self.rate.sleep()
        ## Reset the angular rotation value to be zero
        self.twist.angular.z = 0.0
        self.twist.linear.x = 0.0

    def left_turn(self, iter):
        ## Set angular rotation around z access to 1 (turn left/CCW)
        self.twist.angular.z = 1.0

        for i in range(iter):
            self.pub.publish(self.twist)
            self.rate.sleep()
        ## Reset the angular rotation value to be zero
        self.twist.angular.z = 0.0


    def right_turn(self, iter):
        ## Set angular rotation around z access to -1 (turn right/CW)
        self.twist.angular.z = -.90

        for i in range(iter):
            self.pub.publish(self.twist)
            self.rate.sleep()
        ## Reset the angular rotation value to be zero
        self.twist.angular.z = 0.0

    def move_forward(self,iter):
        ## Set angular rotation around z access to -1 (turn right/CW)
        self.twist.linear.x = 0.2
        for i in range(iter):
            self.pub.publish(self.twist)
            self.rate.sleep()

        ## Reset the angular rotation value to be zero
        self.twist.linear.x = 0.0

    def move_backward(self,iter):
        ## Set angular rotation around z access to -1 (turn right/CW)
        self.twist.linear.x = -0.2
        for i in range(iter):
            self.pub.publish(self.twist)
            self.rate.sleep()

        ## Reset the angular rotation value to be zero
        self.twist.linear.x = 0.0



if __name__ == "__main__":
    ## Create a node
    rospy.init_node("la_chona_dance", anonymous = False)

    ## Make sure sim time is working
    while not rospy.Time.now():
        pass

    ## Setup clients
    arm_action = FollowTrajectoryClient()
    head_action = PointHeadClient()
    base_action = FootWork()


    ## init configuration
    rospy.sleep(.5)
    head_action.look_at(1.0, 0.0, 1.2, duration = 1)
    arm_action.safe_move_to([0.3, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], velocity = .5)
    rospy.sleep(1)

    ## 4 basics moving forward
    for i in range(4):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Forward")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Forward")

    ## 4 basics moving forward
    for i in range(4):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Backward")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Backward")

    ## 2 loops of right and left turns
    for i in range(2):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Right_turn")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Right_turn")
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Left_turn")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Left_turn")

    ## 2 basics turning right
    for i in range(2):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Right_turn")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Right_turn")

    ## 2 basics moving forward
    for i in range(2):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Forward")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Forward")

    ## 2 basics turning left
    for i in range(3):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Left_turn")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Left_turn")

    ## 2 basics moving forward
    for i in range(3):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Forward")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Forward")

    ## 2 basics turning right
    for i in range(2):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Right_turn")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Right_turn")

    ## 3 basics moving backward
    for i in range(2):
        arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Backward")
        arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4, base_motion = "Backward")


    rospy.sleep(.4)

    ## Only 3 movements for torso. Goes better with song
    arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4)
    arm_action.fast_move_to([0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4)
    arm_action.fast_move_to([0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0], duration = 0.4)

    ## Begin wide left turn while changing new arm configuration
    arm_action.fast_move_to([0.3, -1.33, 0.21, 2.75, 1.84, 0.0, 1.03, 0.0], duration = 4.0, base_motion = "Wide_turn")

    rospy.sleep(.4)

    ## 7 basics turning left
    for i in range(7):
        arm_action.fast_move_to([0.28, -1.33, 0.21, 2.75, 1.84, 0.0, 1.03, 0.0], duration = 0.4, base_motion = "Left_turn")
        arm_action.fast_move_to([0.3, -1.33, 0.21, 2.75, 1.84, 0.0, 1.03, 0.0], duration = 0.4, base_motion = "Left_turn")

    ## 4 basics moving forward
    for i in range(4):
        arm_action.fast_move_to([0.28, -1.33, 0.21, 2.75, 1.84, 0.0, 1.03, 0.0], duration = 0.4, base_motion = "Forward")
        arm_action.fast_move_to([0.3, -1.33, 0.21, 2.75, 1.84, 0.0, 1.03, 0.0], duration = 0.4, base_motion = "Forward")

    ## Grab sobrero and remove
    arm_action.fast_move_to([0.3, -1.33, 0.04, 2.64, 1.94, 0.0, 1.03, 0.69], duration = 1)
    arm_action.close_gripper()
    arm_action.fast_move_to([0.3, -1.31, -0.29, 2.64, 1.33, 0.0, 0.59, 0.69], duration = 1)

    rospy.sleep(5)
    arm_action.open_gripper()

File: src/la_chona_dance.cpp
#include <ros/ros.h>
#include "PointHeadClient.h"
#include "Footwork.h"
#include "FollowTrajectoryClient.h"
#include "GripperClient.h"
#include "MoveItClient.h"

/**
 * @brief Entry point of the `la_chona_dance` program.
 *
 * @param argc Number of command-line arguments passed to the program.
 * @param argv Array of command-line arguments.
 *
 * @return int Returns 0 upon successful execution.
 */
int main(int argc, char** argv) {
    // Initialize the ROS node
    ros::init(argc, argv, "la_chona_dance");

    // // Create instances of `PointHeadClient`. This class provides functionality to interact
    // // with the robot's head controllers through ROS action servers.
    PointHeadClient head_client;

    // // Create an instance of `GripperClient`. This class allows control over the robot's gripper,
    // // enabling commands to open and close the gripper for object manipulation.
    GripperClient gripper_client;

    // // Create an instance of `Footwork`. This class provides functionality for base movements,
    // // such as linear motion and turning. It sends velocity commands to the robot's base.
    Footwork base_cmd;

    // // Create an instance of `FollowTrajectoryClient`. This class interacts with the
    // // robot's arm, allowing it to move to specified joint positions through a trajectory action.
    FollowTrajectoryClient trajectory_client;

    // // Initialize the MoveGroupClient to interface with MoveIt for motion planning and execution.
    MoveGroupClient move_group_client;

    // // Move to initial dance pose using MoveIt to avoid collisions
    head_client.lookAt({1.0, 0.0, 1.2});
    gripper_client.close_gripper(false);
    move_group_client.init_pose({0.3, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.5);

    // // Implement a delay before the start of the dance routine. 
    ros::Duration(5.0).sleep();

    // // 8 basics moving forward and backward, while also moving the torso up and down
    for (int i=0; i<4; i++){
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Forward"); // duration=0.3, base_motion="Forward"
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Forward");
    }

    for (int i=0; i<4; i++){
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Backward"); // duration=0.3, base_motion="Backward"
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Backward");
    }

    // // Small cumbia back steps
    for (int i=0; i<2; i++){
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Right Turn"); // duration=0.3, base_motion="Right Turn"
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Right Turn");
        ros::Duration(0.1).sleep(); // make sure there is a pause before rotating in the other direction
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Left Turn"); // duration=0.3, base_motion="Right Turn"
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Left Turn");
        ros::Duration(0.1).sleep();
    }

    // // Turn to the right and face the camera in a 45 degree angle, them move forward
    trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Right Turn");
    trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Right Turn");

    for (int i=0; i<3; i++){
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Forward");
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Forward");
    }

    // // Turn to the left and face the camera in a 45 degree angle, then move forward
    ros::Duration(0.1).sleep();
    for (int i=0; i<2; i++){
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.29, "Left Turn");
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.29, "Left Turn");
    }
   
    for (int i=0; i<3; i++){
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Forward");
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Forward");
    }

    // // Right turn to face the camera, then move backward.
    ros::Duration(0.1).sleep();
    trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Right Turn");
    trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Right Turn");

    for (int i=0; i<4; i++){
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Backward");
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3, "Backward");
    }

    // // Pause before the drop
    ros::Duration(.5).sleep();

    // // Move torso with tempo
    for (int i=0; i<2; i++){
        trajectory_client.move_joints_to({0.30, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3);
        trajectory_client.move_joints_to({0.28, -1.3, 1.2, 0.29, 1.86, -0.02, 1.29, 0.0}, 0.3);
    }
    ros::Duration(.2).sleep();

    // // Complete a wide right turn while moving the arm configuration
    base_cmd.turn(true, 35);
    trajectory_client.move_joints_to({0.3,  -0.2, -0.52, 3.11, 1.23, -0.03, 1.2, 0.02}, 2.9, "Right Turn"); //

    // // // Left turn to face the camera then move backward
    // for (int i=0; i<4; i++){
    //     trajectory_client.move_joints_to({0.28, -0.2, -0.52, 3.11, 1.23, -0.03, 1.2, 0.02}, 0.251, "Left Turn");
    //     trajectory_client.move_joints_to({0.30, -0.2, -0.52, 3.11, 1.23, -0.03, 1.2, 0.02}, 0.25, "Left Turn");
    // }

    // for (int i=0; i<4; i++){
    //     trajectory_client.move_joints_to({0.28, -1.33, 0.21, 2.75, 1.84, 0.0, 1.03, 0.0}, 0.4, "Backward");
    //     trajectory_client.move_joints_to({0.30, -1.33, 0.21, 2.75, 1.84, 0.0, 1.03, 0.0}, 0.4, "Backward");
    // }

    // // Move the arm to grasp the sombrero then pull it off Fetch's head.
    trajectory_client.move_joints_to({0.3, -0.2, -0.31, 3.11, 1.42, -0.03, 1.45, 0.01}, 1);
    gripper_client.close_gripper(true);
    trajectory_client.move_joints_to({0.3, -1.31, -0.29, 2.64, 1.33, 0.0, 0.59, 0.69}, 1);

    // // Open gripper to restart dance routine
    ros::Duration(5).sleep();
    gripper_client.close_gripper(false);

    // // Return 0 to indicate successful execution of the program.
    return 0;
}
